### ISSUE TITLE
Tech plugins should be able to override hooks

### ISSUE THREAD

**Issue by edwardcwang** [ucb-bar/hammer#311]
e.g. <mytech> plugin should be able to provide hooks in a programmatic way

**Comment by edwardcwang** [comment]
Required: #281

**Comment by jwright6323** [comment]
Basically want the priority to be:
1 core hammer steps
2 technology hooks
3 user-supplied hooks

**Comment by edwardcwang** [comment]
We want priority and yet at the same time if the user desires, the user should be able to override core/tech steps with custom steps yes?

**Comment by jwright6323** [comment]
Yes, 2 can overwrite/remove/add to 1, and 3 can overwrite/remove/add to 2 or 1.

**Comment by harrisonliew** [comment]
Question of nuances here. Is the desired overwriting behavior:
1. Given core step `foo` and tech-supplied `make_pre_insertion_hook(foo, bar)`, user should supply `make_replacement_hook(bar)` to override `bar`, or
2. Given core step `foo`, user-supplied `make_pre_insertion_hook(foo, bar)` will override tech-supplied `make_pre_insertion_hook(foo, bar)`?

If 1. is desired, the existing code works as long as hooks are ordered w.r.t. priority.
If 2. is desired, it's a bit more code, and conflicts with this interpretation of "adding" to steps/hooks which is also enabled by 1.:
Given core step `foo`, tech & user supply `make_pre_insertion_hook(foo, bar)`. Tech's `bar` is executed followed by user's `bar` followed by `foo`.

Any other use cases I'm missing here?

**Comment by jwright6323** [comment]
1 is what I was expecting. 2 sounds difficult to implement and hard to follow when debugging. I think 1 allows you to do all possible things:
1. add a tech hook
2. overwrite a tech hook
3. remove a tech hook
4. add non-tech hooks
5. overwrite non-tech steps
6. remove non-tech steps
